src/hpa.c

```diff
@@ -0,0 +1,395 @@
+	 * overcommitting systems, and we're not going to purge much.
+	 */
+	bool commit = true;
+	void *addr = pages_map(NULL, alloc_size, HUGEPAGE, &commit);
+	if (addr == NULL) {
+		malloc_mutex_unlock(tsdn, &hpa->grow_mtx);
+		return NULL;
+	}
+	err = pages_huge(addr, alloc_size);
+	/*
+	 * Ignore this for now; even if the allocation fails, the address space
+	 * should still be usable.
+	 */
+	(void)err;
+
+	edata = edata_cache_get(tsdn, hpa->edata_cache);
+	if (edata == NULL) {
+		malloc_mutex_unlock(tsdn, &hpa->grow_mtx);
+		pages_unmap(addr, alloc_size);
+		return NULL;
+	}
+
+	edata_init(edata, hpa->ind, addr, alloc_size, /* slab */ false,
+	    SC_NSIZES, /* sn */ 0, extent_state_active, /* zeroed */ true,
+	    /* comitted */ true, EXTENT_PAI_HPA, /* is_head */ true);
+
+	malloc_mutex_lock(tsdn, &hpa->mtx);
+	err = hpa_central_alloc_grow(tsdn, &hpa->central, size_goal, edata);
+	malloc_mutex_unlock(tsdn, &hpa->mtx);
+
+	if (!err) {
+		geom_grow_size_commit(&hpa->geom_grow, skip);
+	}
+	malloc_mutex_unlock(tsdn, &hpa->grow_mtx);
+	edata_arena_ind_set(edata, shard->ind);
+
+	if (err) {
+		pages_unmap(addr, alloc_size);
+		edata_cache_put(tsdn, hpa->edata_cache, edata);
+		return NULL;
+	}
+
+	return edata;
+}
+
+static edata_t *
+hpa_alloc_psset(tsdn_t *tsdn, hpa_shard_t *shard, size_t size) {
+	assert(size < shard->ps_alloc_max);
+
+	bool err;
+	edata_t *edata = edata_cache_get(tsdn, shard->edata_cache);
+	edata_arena_ind_set(edata, shard->ind);
+
+	malloc_mutex_lock(tsdn, &shard->mtx);
+	err = psset_alloc_reuse(&shard->psset, edata, size);
+	malloc_mutex_unlock(tsdn, &shard->mtx);
+	if (!err) {
+		return edata;
+	}
+	/* Nothing in the psset works; we have to grow it. */
+	malloc_mutex_lock(tsdn, &shard->grow_mtx);
+
+	/* As above; check for grow races. */
+	malloc_mutex_lock(tsdn, &shard->mtx);
+	err = psset_alloc_reuse(&shard->psset, edata, size);
+	malloc_mutex_unlock(tsdn, &shard->mtx);
+	if (!err) {
+		malloc_mutex_unlock(tsdn, &shard->grow_mtx);
+		return edata;
+	}
+
+	edata_t *grow_edata = hpa_alloc_central(tsdn, shard, size,
+	    shard->ps_goal);
+	if (grow_edata == NULL) {
+		malloc_mutex_unlock(tsdn, &shard->grow_mtx);
+		edata_cache_put(tsdn, shard->edata_cache, edata);
+		return NULL;
+	}
+	edata_arena_ind_set(grow_edata, shard->ind);
+	edata_slab_set(grow_edata, true);
+	fb_group_t *fb = edata_slab_data_get(grow_edata)->bitmap;
+	fb_init(fb, shard->ps_goal / PAGE);
+
+	/* We got the new edata; allocate from it. */
+	malloc_mutex_lock(tsdn, &shard->mtx);
+	psset_alloc_new(&shard->psset, grow_edata, edata, size);
+	malloc_mutex_unlock(tsdn, &shard->mtx);
+
+	malloc_mutex_unlock(tsdn, &shard->grow_mtx);
+	return edata;
+}
+
+static edata_t *
+hpa_alloc(tsdn_t *tsdn, pai_t *self, size_t size,
+    size_t alignment, bool zero) {
+
+	assert((size & PAGE_MASK) == 0);
+	/* We don't handle alignment or zeroing for now. */
+	if (alignment > PAGE) {
+		return NULL;
+	}
+	if (zero) {
```

Comment:
nit: slightly prefer combining the two branches. feel free to ignore.

Reply:
Agreed, done.